### PR TITLE
[WIP] neighbourhood addressing for unencrypted pss

### PR DESCRIPTION
**Pull request by zelig**
This PR introduces neighbourhood addressing for unencrypted PSS messaging.
It is WIP, missing tests!!!

Neighbourhood is addressed by specifying a destination address and either or both of the following:

* minimum neighbourhood radius `R`: 
   self is regarded as possible recipient iff they share at least R prefix bits
   handler is called if `PO(local, destination) >= R` 
* neighbourhood size `S`:
   depth for `S` `D(S)` is the highest proximity order such that the node is connected to at least S peers.
   self regarded as possible recipient iff `PO(local, destination) >= D(S)`

These addressing modes can be combined with encryption in which case they effectively implement an obfuscated version of dark messaging.

The primary usecase though is unencrypted (raw) messaging to implement PUSH syncing, i.e. sending chunks to their redundancy neighbourhood.

### CODE REVIEW

File: swarm/pss/pss.go [hunk 1]
```diff
@@ -430,7 +430,9 @@ func (p *Pss) process(pssmsg *PssMsg) error {
 
 func (p *Pss) executeHandlers(topic Topic, payload []byte, from *PssAddress, asymmetric bool, keyid string) {
 	handlers := p.getHandlers(topic)
-	peer := p2p.NewPeer(enode.ID{}, fmt.Sprintf("%x", from), []p2p.Cap{})
+	// nid := enode.HexID("0x00") // this hack is needed to satisfy the p2p method
+	nid := enode.ID{}
+	peer := p2p.NewPeer(nid, fmt.Sprintf("%x", from), []p2p.Cap{})
```
Comment (nolash) [inline swarm/pss/pss.go:435]: ?

File: swarm/pss/types.go [hunk 2]
```diff
@@ -91,29 +93,65 @@ type pssDigest [digestLength]byte
 
 // conceals bitwise operations on the control flags byte
 type msgParams struct {
-	raw bool
-	sym bool
+	raw bool // unenctypted mode
+	sym bool // symmetric encryption
+	nhr int  // neighbourhood addressing mode based on minumim proximity order
```
Comment (nolash) [inline swarm/pss/types.go:98]: Why `int`? Doesn't `byte` suffice?

File: swarm/pss/pss.go [hunk 3]
```diff
@@ -446,8 +448,25 @@ func (p *Pss) isSelfRecipient(msg *PssMsg) bool {
...
+	params := newMsgParamsFromBytes(msg.Control)
 	local := p.Kademlia.BaseAddr()
-	return bytes.Equal(msg.To, local[:len(msg.To)])
+	if params.nhr == 0 && params.nhs == 0 {
+		return bytes.Equal(msg.To, local[:len(msg.To)])
+	}
+	minProx := 255
+	if params.nhr > 0 {
+		minProx = params.nhr
+	}
```
Comment (nolash) [inline swarm/pss/pss.go:461]: I prefer `if ... else` if you don't mind

File: swarm/pss/types.go [hunk 4]
```diff
@@ -91,29 +93,65 @@ type pssDigest [digestLength]byte
...
+		} else {
+			return nil
+		}
+	}
+
 	return &msgParams{
 		raw: paramBytes[0]&pssControlRaw > 0,
 		sym: paramBytes[0]&pssControlSym > 0,
+		nhs: nhs,
+		nhr: nhr,
```
Comment (nolash) [inline swarm/pss/types.go:132]: Why do we need both in the bitfield? Will one ever be set without the other?

File: swarm/pss/types.go [hunk 5]
```diff
@@ -91,29 +93,65 @@ type pssDigest [digestLength]byte
...
+	if m.nhr > 0 {
+		b |= pssControlNeighbourhoodRadius
+		nh = append(nh, byte(m.nhr))
+	}
+	if m.nhs > 0 {
+		b |= pssControlNeighbourhoodSize
+		nh = append(nh, byte(m.nhs))
+	}
 	paramBytes = append(paramBytes, b)
+	paramBytes = append(paramBytes, nh...)
```
Comment (nolash) [inline swarm/pss/types.go:154]: Will empty byte here still cause serialization of "empty byte" in `rlp` do you know?

File: swarm/pss/pss.go [hunk 6]
```diff
@@ -446,8 +448,25 @@ func (p *Pss) isSelfRecipient(msg *PssMsg) bool {
 
 // test match of leftmost bytes in given message to node's Kademlia address
 func (p *Pss) isSelfPossibleRecipient(msg *PssMsg) bool {
+	// in partial addressing mode self is possible recipient if To address is a prefix
+	// of the local kademlia base address
+	params := newMsgParamsFromBytes(msg.Control)
```
Comment (nolash) [inline swarm/pss/pss.go:453]: Please let's not parse the struct before we check the bits.

**Review by nolash — CHANGES_REQUESTED** [review]

**Comment by nolash** [comment]
> These addressing modes can be combined with encryption in which case they effectively implement an obfuscated version of dark messaging.

I'm not sure what this means.

**Comment by nolash** [comment]
> minimum neighbourhood radius R:
self is regarded as possible recipient iff they share at least R prefix bits
handler is called if PO(local, destination) >= R
neighbourhood size S:
depth for S D(S) is the highest proximity order such that the node is connected to at least S peers.
self regarded as possible recipient iff PO(local, destination) >= D(S)

This means that if I'm connected to at least S peers that have the same prefix D as me, then it's for me?

File: swarm/pss/pss.go (hunk 6 above)
Comment (zelig) [reply swarm/pss/pss.go:453]: why? i dont get it

File: swarm/pss/types.go (hunk 4 above)
Comment (zelig) [reply swarm/pss/types.go:132]: yes why not

File: swarm/pss/types.go (hunk 5 above)
Comment (zelig) [reply swarm/pss/types.go:154]: No, i dont understand the question :(

File: swarm/pss/types.go (hunk 4 above)
Comment (nolash) [reply swarm/pss/types.go]: in `isSelfPossibleReceipient` nhr is ever only used if nhs is set.

File: swarm/pss/types.go (hunk 2 above)
Comment (nolash) [reply swarm/pss/types.go:98]: i don't understand. Can you have proximity orders > 256?

File: swarm/pss/types.go (hunk 5 above)
Comment (nolash) [reply swarm/pss/types.go:154]: If `nh == []byte{}` will the rlp serialization be longer if `nh` is appended than if it isn't?

File: swarm/pss/pss.go (hunk 6 above)
Comment (nolash) [reply swarm/pss/pss.go:453]: Never mind, the actual content doesn't get parsed in the `newMsgParamsFromBytes` if they're not set.

**Comment by nolash** [comment]
I am thinking; wouldn't merely a reserved topic `0x00` or merely just a simple `syncpacket` flag and using the implementation specific settings already existing for proxbins/neighborhoods/syncing make this added complexity unnecessary?

Can we please look at the other benefits of this system, other than actual syncing?

File: swarm/pss/types.go (hunk 4 above)
Comment (nolash) [reply swarm/pss/types.go:132]: In which cases would you only specify S and partial `To` address for example?

File: swarm/pss/types.go (hunk 2 above)
Comment (zelig) [reply swarm/pss/types.go:98]: ok it could be uint8 but then I need to convert elsewhere

File: swarm/pss/types.go (hunk 4 above)
Comment (zelig) [reply swarm/pss/types.go]: with S specified you can afford to have a partial address only. 
What i mean by obfuscated dark message is that sending to a n-byte partial address is equivalent to sending to a bogus full address with at least n-long full address and specifying R=8n. which implies that partial addressing can be eliminated really.

Comment (zelig) [reply swarm/pss/types.go:132]: cannot think of a use case for that. In fact if the partial TO address is shorter than the Depth needed for S then noone will pick up the message...

File: swarm/pss/types.go (hunk 5 above)
Comment (zelig) [reply swarm/pss/types.go:154]: apoending an the empty byte slice is doing nothing so serialisation does not change either

**Comment by zelig** [comment]
obsoleted in favour of  #965

**Pull request closed without merging**